src/binder/clrprivbinderassemblyloadcontext.cpp

```diff
@@ -268,9 +299,80 @@ HRESULT CLRPrivBinderAssemblyLoadContext::SetupContext(DWORD      dwAppDomainId,
     return hr;
 }
 
+#ifdef FEATURE_COLLECTIBLE_ALC
+
+/* static */
+BOOL CLRPrivBinderAssemblyLoadContext::DestroyContext(CLRPrivBinderAssemblyLoadContext *pBindContext)
+{
+    {
+        CrstHolder holder(pBindContext->m_loadersCrst);
+
+        LoaderAllocatorSet &loaderAllocators = pBindContext->m_loaderAllocators;
+
+        LoaderAllocatorSet::Iterator iter = loaderAllocators.Begin();
+        while (iter != loaderAllocators.End())
+        {
+            LoaderAllocator *pLoaderAllocator = *iter;
+
+            // This context should be holding onto a reference to the LoaderAllocator, so it
+            // should still be alive
+            _ASSERTE(pLoaderAllocator->IsAlive());
+
+            if (pLoaderAllocator->IsManagedScoutAlive())
+            {
+                // We can't destroy until there is no managed reference to any of our LoaderAllocators
+                return FALSE;
+            }
+
+            iter++;
+        }
+
+        iter = loaderAllocators.Begin();
+        while (iter != loaderAllocators.End())
+        {
+            LoaderAllocator *pLoaderAllocator = *iter;
+
+            pLoaderAllocator->ReleaseAllReferences();
+
+            // Release our reference to the LoaderAllocator so it can be deleted by the LoaderAllocator GC
+            VERIFY(pLoaderAllocator->Release());
+
+            iter++;
+        }
+    }
+
+    // TODO: Is AppDomain::GetCurrentDomain() going to work properly?
+    AppDomain *pAppDomain = AppDomain::GetCurrentDomain();
+
+    // The managed LoaderAllocatorScout finalizer would normally trigger the LoaderAllocator GC
+    // but because we hold a reference that outlasts it, we need to trigger it as well
+    LoaderAllocator::GCLoaderAllocators(pAppDomain);
+    
+    // The managed AssemblyLoadContext should be the only remaining reference to the native
+    // binding context
+    VERIFY(pBindContext->Release() == 0);
+
+    return TRUE;
+}
+
+#endif // FEATURE_COLLECTIBLE_ALC
+
 CLRPrivBinderAssemblyLoadContext::CLRPrivBinderAssemblyLoadContext()
 {
     m_pTPABinder = NULL;
+
+#ifdef FEATURE_COLLECTIBLE_ALC
+    m_isCollectible = FALSE;
+
+    m_loadersCrst = new Crst(CrstType::CrstAssemblyLoadContextLoaderAllocators);
```

Comment:
I'm not sure how to do that, Crst overrides new which is preventing me from using nothrow.